x-pack/plugins/security/server/plugin.ts

```diff
@@ -141,6 +141,12 @@ export class Plugin {
       .pipe(first())
       .toPromise();
 
+    // A subset of `start` services we need during `setup`.
+    const startServicesPromise = core.getStartServices().then(([coreServices, depsServices]) => ({
+      elasticsearch: coreServices.elasticsearch,
+      features: depsServices.features,
+    }));
```

Comment:
I have no idea yet how we're going to solve this but I had to do the same for audit logging and it's a bit of an architectural smell. Lifecycles should flow in one direction. A lifecycle can certainly have access to a previous phase but not the other way around. If we need access to a start contract in the setup phase then either the setup contract we're depending on isn't exposing everything it should or the lifecycles themselves aren't really separate lifecycles. 

Would love to start thinking about improving this. (Obviously not as part of this PR)

@elastic/kibana-core Are you guys aware this is happening throughout the codebase? Any ideas or plans how this situation could be avoided in the future?

Reply:
> it's a bit of an architectural smell. 

It's! And one of the main reasons why we still use this "trick", in Security at least, is that we need all of that stuff in the HTTP routes and we can only register them at the `setup` stage right now. Would definitely love to get rid of this.

Reply:
> Are you guys aware this is happening throughout the codebase? 

We are.

> Any ideas or plans how this situation could be avoided in the future?

Using context-based wrapper around all APIs that might require start contract when registered in setup would be a solution. (such as what is done with `RouteHandlerContext` for http handler for example). This is some heavy code restructuring though, and is not planned atm.